app/commands/track/determine_concepts_layout.rb

```diff
@@ -6,96 +6,115 @@ class Track
-        node.concepts.each { |concept| lookup[concept] = node }
-      end.freeze
-    end
+    class Graph
+      include Mandate
 
-    def make_index_lookup
-      nodes.index_by(&:index).freeze
-    end
+      # Edge representing a directed edge
+      Edge = Struct.new(:from, :to, keyword_init: true)
 
-    def make_edges
-      nodes.each_with_object([]) do |node, aggregate|
-        node.prerequisites.each_with_object(aggregate) do |prereq, a|
-          prereq_node = node_lookup_by_concept[prereq]
-          a << Edge.new(from: prereq_node, to: node) unless prereq_node.nil?
-        end
-      end.freeze
-    end
+      # Node for representing an exercise within a track
+      Node = Struct.new(:index, :slug, :uuid, :concepts, :prerequisites, keyword_init: true)
 
-    #
-    # Supporting Graph Classes
-    #
+      # Add public readers
+      attr_reader :nodes, :edges
 
-    # Edge representing a directed edge
-    class Edge
-      attr_reader :from, :to
+      def initialize(exercises)
+        @exercises = exercises
+        @nodes = determine_nodes
+        @edges = determine_edges
+      end
 
-      def initialize(from:, to:)
-        @from = from
-        @to = to
+      memoize
+      def empty?
+        nodes.empty?
+      end
+
+      memoize
+      def has_cycle?
+        GraphUtils.contains_cycle?(adjacencies)
+      end
+
+      def node_for_concept(concept)
+        nodes_keyed_by_concept[concept]
       end
-    end
 
-    # Node for representing an exercise within a track
-    class Node
-      attr_reader :index, :slug, :concepts, :prerequisites
+      def node_for_index(index)
+        nodes_keyed_by_index[index]
+      end
+
+      memoize
+      def levels
+        GraphUtils.calculate_levels(adjacencies)
+      end
+
+      private
+      attr_reader :exercises
+
+      # Creates adjacency list for a graph with directed edges
+      memoize
+      def adjacencies
+        Array.new(nodes.length) { [] }.tap do |adjacencies|
+          edges.each do |edge|
+            from = edge.from.index
+            to = edge.to.index
+            adjacencies[from] << to
+          end
+        end
+      end
+
+      def determine_nodes
+        exercises.map.with_index do |exercise, idx|
+          Node.new(
+            index: idx,
+            slug: exercise['slug'],
+            uuid: exercise['uuid'],
+            concepts: exercise['concepts'],
+            prerequisites: exercise['prerequisites']
+          )
+        end.freeze
+      end
+
+      def determine_edges
+        nodes.flat_map do |node|
+          node.prerequisites.map do |prereq|
+            prereq_node = node_for_concept(prereq)
+            Edge.new(from: prereq_node, to: node) unless prereq_node.nil?
+          end
+        end.freeze
+      end
+
+      memoize
+      def nodes_keyed_by_concept
+        nodes.each_with_object({}) do |node, lookup|
+          node.concepts.each { |concept| lookup[concept] = node }
+        end.freeze
```

Comment:
```suggestion
        nodes.flat_map do |node|
          node.concepts.map { |concept| [concept, node] }
        end.to_h.freeze
```

This is another way of doing the same thing. I'm not sure which I prefer 🙂

Reply:
I like your suggested way myself now that I see it

Reply:
The reason I'm not sure here is because your version is very readable. If there were multiple LOC within either of the blocks, I'd definitely prefer the later I think (less passing things around etc) but it's very clear what your code does and potentially more clear than my version here.

Reply:
Is there a way to declare the intention to turn it into a hash before the flatmap? maybe it seems less clear because the desired end-format isn't clear from the start?